File: java/src/main/java/com/workshop/architecture/fitness/clean_architecture/use_case/ActivateMembershipUseCaseInteractor.java
package com.workshop.architecture.fitness.clean_architecture.use_case;

import com.workshop.architecture.fitness.clean_architecture.entity.*;
import com.workshop.architecture.fitness.clean_architecture.use_case.exception.CustomerTooYoungException;
import com.workshop.architecture.fitness.clean_architecture.use_case.port.inbound.ActivateMembershipUseCase;
import com.workshop.architecture.fitness.clean_architecture.use_case.port.inbound.ActivateMembershipInput;
import com.workshop.architecture.fitness.clean_architecture.use_case.port.inbound.ActivateMembershipResult;
import com.workshop.architecture.fitness.clean_architecture.use_case.port.outbound.*;
import jakarta.transaction.Transactional;
import org.jspecify.annotations.NonNull;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.time.Instant;
import java.time.LocalDate;
import java.util.UUID;

import static java.lang.Boolean.TRUE;

@Service
public class ActivateMembershipUseCaseInteractor implements ActivateMembershipUseCase {
    private final ForFindingCustomers forFindingCustomers;
    private final ForFindingPlans forFindingPlans;
    private final ForStoringMemberships forStoringMemberships;
    private final ForCreatingInvoices forCreatingInvoices;
    private final ForStoringBillingReferences forStoringBillingReferences;
    private final ForSendingEmails forSendingEmails;

    public ActivateMembershipUseCaseInteractor(
            ForFindingCustomers forFindingCustomers,
            ForFindingPlans forFindingPlans,
            ForStoringMemberships forStoringMemberships,
            ForCreatingInvoices forCreatingInvoices,
            ForStoringBillingReferences forStoringBillingReferences,
            ForSendingEmails forSendingEmails
    ) {
        this.forFindingCustomers = forFindingCustomers;
        this.forFindingPlans = forFindingPlans;
        this.forStoringMemberships = forStoringMemberships;
        this.forCreatingInvoices = forCreatingInvoices;
        this.forStoringBillingReferences = forStoringBillingReferences;
        this.forSendingEmails = forSendingEmails;
    }

    @Transactional
    @Override
    public ActivateMembershipResult activateMembership(ActivateMembershipInput input) {

        var customerId = UUID.fromString(input.customerId());
        var customer = forFindingCustomers.findCustomerById(customerId);

        var planId = UUID.fromString(input.planId());
        var plan = forFindingPlans.findPlanById(planId);

        var membership = createMembership(customer, customerId, planId, input.signedByCustodian(), LocalDate.now(), plan.durationInMonths(), plan.price());

        var storedMembership = forStoringMemberships.storeMembership(membership);
        var invoiceDetails = new InvoiceDetails(
                UUID.randomUUID(),
                customerId,
                membership.id(),
                plan.id(),
                LocalDate.now().plusDays(30),
                plan.title(),
                plan.price());

        var externalInvoiceId = forCreatingInvoices.createInvoiceWith(invoiceDetails);


        var now = Instant.now();
        var billingReference = new MembershipBillingReference(
                UUID.randomUUID(),
                storedMembership.id(),
                externalInvoiceId,
                invoiceDetails.membershipId().toString(),
                invoiceDetails.dueDate(),
                "OPEN",
                now,
                now
        );

        var storedBillingReference = forStoringBillingReferences.storeMembershipBillingReference(billingReference);

        forSendingEmails.sendEmail(toEmail(storedBillingReference, customer.emailAddress(), storedMembership.planPrice()));

        return new ActivateMembershipResult(
                storedBillingReference.membershipId().toString(),
                storedMembership.customerId().toString(),
                storedMembership.planId().toString(),
                storedMembership.planPrice(),
                storedMembership.planDurationInMonths(),
                storedMembership.status(),
                storedMembership.startDate(),
                storedMembership.endDate(),
                storedBillingReference.externalInvoiceReference(),
                storedBillingReference.externalInvoiceId(),
                storedBillingReference.dueDate()
        );
    }

    private static Membership createMembership(Customer customer, UUID customerId, UUID planId, Boolean isSignedByCustodian, LocalDate startDate, int planDurationInMonths, BigDecimal planPrice) {
        var endDate = startDate.plusMonths(planDurationInMonths);

        if (!customer.isAdultAt(startDate) && isMissingSignature(isSignedByCustodian)) {
            throw new CustomerTooYoungException(
                    "Customers younger than 18 require signedByCustodian=true"
            );
        }

        return new Membership(
                UUID.randomUUID(),
                customerId,
                planId,
                planPrice.intValue(),
                planDurationInMonths,
                "ACTIVE",
                null,
                startDate,
                endDate
        );
    }

    private static boolean isMissingSignature(Boolean isSignedByCustodian) {
        return !TRUE.equals(isSignedByCustodian);
    }


    private static @NonNull CustomerActivateMembershipEmail toEmail(MembershipBillingReference billingReference, String emailAddress, int membershipPlanPrice) {
        return new CustomerActivateMembershipEmail(
                billingReference.externalInvoiceReference(),
                billingReference.dueDate(),
                emailAddress,
                membershipPlanPrice,
                """
                        |
                        |To: %s
                        |From: %s
                        |Subject: Your Membership Invoice %s
                        |
                        |Dear customer,
                        |
                        |Thank you for your membership.
                        |
                        |Please find your invoice details below:
                        |Invoice ID: %s
                        |Amount Due: CHF %s
                        |Due Date: %s
                        |
                        |Attachment: invoice-%s.pdf
                        |
                        |Kind regards,
                        |Codeartify Billing
                        |
                        """);
    }

}

File: python/src/workshop_api/fitness/clean_architecture/use_case/activate_membership_use_case_interactor.py
import uuid
from datetime import date

from ..entity import (
    CustomerActivateMembershipEmail,
    Membership,
    MembershipBillingReference,
    MembershipInvoiceDetails,
)
from .port.inbound import (
    ActivateMembershipInput,
    ActivateMembershipResult,
    ActivateMembershipUseCase,
)
from .port.outbound import (
    ForCreatingInvoices,
    ForFindingCustomers,
    ForFindingPlans,
    ForSendingEmails,
    ForStoringBillingReferences,
    ForStoringMemberships,
)


class ActivateMembershipUseCaseInteractor(ActivateMembershipUseCase):
    def __init__(
        self,
        for_finding_customers: ForFindingCustomers,
        for_finding_plans: ForFindingPlans,
        for_storing_memberships: ForStoringMemberships,
        for_creating_invoices: ForCreatingInvoices,
        for_storing_billing_references: ForStoringBillingReferences,
        for_sending_emails: ForSendingEmails,
    ) -> None:
        self.for_finding_customers = for_finding_customers
        self.for_finding_plans = for_finding_plans
        self.for_storing_memberships = for_storing_memberships
        self.for_creating_invoices = for_creating_invoices
        self.for_storing_billing_references = for_storing_billing_references
        self.for_sending_emails = for_sending_emails

    async def activate_membership(
        self,
        input_data: ActivateMembershipInput,
    ) -> ActivateMembershipResult:
        customer_id = uuid.UUID(input_data.customer_id)
        customer = self.for_finding_customers.find_customer_by_id(customer_id)

        plan_id = uuid.UUID(input_data.plan_id)
        plan = self.for_finding_plans.find_plan_by_id(plan_id)

        membership = Membership.create(
            customer,
            plan_id,
            input_data.signed_by_custodian,
            date.today(),
            plan.duration_in_months,
            plan.price,
        )

        stored_membership = self.for_storing_memberships.store_membership(membership)
        invoice_details = MembershipInvoiceDetails.create(customer_id, membership, plan)
        external_invoice_id = await self.for_creating_invoices.create_invoice_with(
            invoice_details
        )
        billing_reference = MembershipBillingReference.create(
            stored_membership,
            external_invoice_id,
            invoice_details,
        )
        stored_billing_reference = (
            self.for_storing_billing_references.store_membership_billing_reference(
                billing_reference
            )
        )

        email = CustomerActivateMembershipEmail.to_email(
            stored_billing_reference,
            customer.email_address,
            stored_membership.plan_price,
        )
        self.for_sending_emails.send_email(email)

        return self._to_result(stored_billing_reference, stored_membership)

    @staticmethod
    def _to_result(
        stored_billing_reference: MembershipBillingReference,
        membership: Membership,
    ) -> ActivateMembershipResult:
        return ActivateMembershipResult(
            membership_id=str(stored_billing_reference.membership_id),
            customer_id=str(membership.customer_id),
            plan_id=str(membership.plan_id),
            plan_price=membership.plan_price,
            plan_duration=membership.plan_duration_in_months,
            status=membership.status,
            start_date=membership.start_date,
            end_date=membership.end_date,
            invoice_id=stored_billing_reference.external_invoice_reference,
            external_invoice_id=stored_billing_reference.external_invoice_id,
            invoice_due_date=stored_billing_reference.due_date,
        )
